library/std/src/sys/uefi/net/tcp4.rs

```diff
@@ -0,0 +1,473 @@
+            .map_err(|_| common::VARIABLE_BOX_ERROR)?;
+
+        // Initialize TransmitData
+        unsafe {
+            let ptr = transmit_data.as_mut_ptr();
+            addr_of_mut!((*ptr).push).write(r_efi::efi::Boolean::TRUE);
+            addr_of_mut!((*ptr).urgent).write(r_efi::efi::Boolean::FALSE);
+            addr_of_mut!((*ptr).data_length).write(buf_size);
+            addr_of_mut!((*ptr).fragment_count).write(1);
+            (*ptr).fragment_table.as_mut_ptr().copy_from([fragment_table].as_ptr(), 1);
+        };
+        let mut transmit_data = unsafe { transmit_data.assume_init() };
+
+        let packet = tcp4::IoTokenPacket { tx_data: transmit_data.as_mut() };
+        let mut transmit_token = tcp4::IoToken { completion_token, packet };
+        unsafe { Self::transmit_raw(self.protocol.as_ptr(), &mut transmit_token) }?;
+
+        transmit_event.wait()?;
+
+        let r = transmit_token.completion_token.status;
+        if r.is_error() {
+            Err(status_to_io_error(r))
+        } else {
+            Ok(unsafe { (*transmit_token.packet.tx_data).data_length } as usize)
+        }
+    }
+
+    pub(crate) fn transmit_vectored(&self, buf: &[IoSlice<'_>]) -> io::Result<usize> {
+        let buf_size = crate::mem::size_of_val(buf);
+        let transmit_event = common::Event::create(
+            r_efi::efi::EVT_NOTIFY_WAIT,
+            r_efi::efi::TPL_CALLBACK,
+            Some(nop_notify4),
+            None,
+        )?;
+        let completion_token =
+            tcp4::CompletionToken { event: transmit_event.as_raw_event(), status: Status::ABORTED };
+        let fragment_tables: Vec<tcp4::FragmentData> = buf
+            .iter()
+            .map(|b| tcp4::FragmentData {
+                fragment_length: crate::mem::size_of_val(b) as u32,
+                fragment_buffer: (*b).as_ptr() as *mut crate::ffi::c_void,
+            })
+            .collect();
+
+        let layout = unsafe {
+            crate::alloc::Layout::from_size_align_unchecked(
+                crate::mem::size_of::<tcp4::TransmitData>()
+                    + crate::mem::size_of_val(&fragment_tables),
+                POOL_ALIGNMENT,
+            )
+        };
+        let mut transmit_data = VariableBox::<tcp4::TransmitData>::try_new_uninit(layout)
+            .map_err(|_| common::VARIABLE_BOX_ERROR)?;
+        let fragment_tables_len = fragment_tables.len();
+
+        // Initialize TransmitData
+        unsafe {
+            let ptr = transmit_data.as_mut_ptr();
+            addr_of_mut!((*ptr).push).write(r_efi::efi::Boolean::TRUE);
+            addr_of_mut!((*ptr).urgent).write(r_efi::efi::Boolean::FALSE);
+            addr_of_mut!((*ptr).data_length).write(buf_size as u32);
+            addr_of_mut!((*ptr).fragment_count).write(fragment_tables_len as u32);
+            (*ptr)
+                .fragment_table
+                .as_mut_ptr()
+                .copy_from(fragment_tables.as_ptr(), fragment_tables_len);
```

Comment:
I don't think this is sound: since `fragment_table` is zero-sized, the provenance of the returned pointer does not include any more bytes.
```suggestion
            addr_of_mut!((*ptr).fragment_table)
                .cast::<tcp4::FragmentData>()
                .copy_from(fragment_tables.as_ptr(), fragment_tables_len);
```